Fix handler leak #925

basilevs opened the pull request:
Fixes #925
Fixes a regression introduced in https://github.com/eclipse-platform/eclipse.platform.ui/pull/1253

Handlers can store references to external objects so cleaning them up when another handler is activated is not quick enough. WeakReference is enough to detect handler switch, so we use that.

The main problem with this fix is the loss of side-effects on cleanup of old instances of `IObjectWithState` as `removeState()` is never called for them.

basilevs committed:
Fix handler leak #925

Fixes #925

Handlers can store references to external objects so cleaning them up
when another handler is activated is not quick enough.
WeakReference is enough to detect handler switch and a cleanup of
collected IObjectWithState is (hopefully) not necessary.

laeubi commented:
> The main problem with this fix is the loss of side-effects on cleanup of old instances of IObjectWithState as removeState() is never called for them.

Is this relevant if the (referenced) object is garbage collected?

basilevs commented:
> > The main problem with this fix is the loss of side-effects on cleanup of old instances of IObjectWithState as removeState() is never called for them.

> Is this relevant if the (referenced) object is garbage collected?

Yes, as side effects may be important.

laeubi commented:
> Yes, as side effects may be important.

Can you explain that? Maybe `IObjectWithState` should have some kind of way to register a `DisposeListener` then instead?

basilevs commented:
> > Yes, as side effects may be important.
> Can you explain that? Maybe `IObjectWithState` should have some kind of way to register a `DisposeListener` then instead?

When State is added, implementor of IObjectWithState may opt to subscribe to another event source, or otherwise allocate resources. The contract of IObjectWithState is that every `addState()` is paired with `removeState()` and listeners and resource allocation have a chance to be reverted in `removeState()`. I've fixed `AbstractHandlerWithState` to handle disposal gracefully (calling `removeState()` if Command Framework have not already done so, but in general such care is not implied by `IObjectWithState` contract (as can be seen in original implementation of `AbstractHandlerWithState` which did not care).

iloveeclipse commented:
I haven't followed what about original issue is, but using weak reference for something that is very important for that feature would mean, after a "gc" this feature may not work as expected ?

basilevs commented:
> I haven't followed what about original issue is, but using weak reference for something that is very important for that feature would mean, after a "gc" this feature may not work as expected ?

Yes, there is a corner case, when `IObjectWithState` may allocate resources in `addState()` and fail to dispose them as `removeState()` is never called. Usually the disposal should be done in `dispose()` so it should not happen often.

The original idea behind the feature will work as intended, as handlers that are collected are usually no longer interested in command state.

laeubi commented:
> Yes, there is a corner case, when IObjectWithState may allocate resources and fail to dispose them in dispose() method, as removeState() is never called.

I think one can assume that dispose has to remove every remaining resources. The paring is only required if the regitrar object itself is removed and one don't want to leak there.

@basilevs please (as a follow up) enhance the javadoc of `IObjectWithState` that disposing has to cleanup any remaining `addState(...)` if you think there is any doubt and at best check all implementation of `IObjectWithState` in the platform.

iloveeclipse commented:
Have you tried your fix with jdt leak tests - are they "green" now?

basilevs commented:
> Have you tried your fix with jdt leak tests - are they "green" now?

Yes, the were red before the fix and became green after. They were hard to launch on Java 17 as many `--add-opens` were missing from pom file (not sure how that works in CI), but I managed.

Missing VM arguments were:

```
--add-opens jdk.management.jfr/jdk.management.jfr=ALL-UNNAMED
--add-opens jdk.jfr/jdk.jfr.internal.management=ALL-UNNAMED
--add-opens jdk.crypto.ec/sun.security.ec=ALL-UNNAMED
```

basilevs commented:
@laeubi 
> I think one can assume that dispose has to remove every remaining resources. The paring is only required if the regitrar object itself is removed and one don't want to leak there.

Unfortunately, there is no `dispose()` in `IObjectWithState`. In particular `org.eclipse.core.commands.Command` also implements `IObjectWithState` but should not have `dispose()` method or its equivalent.

> @basilevs please (as a follow up) enhance the javadoc of `IObjectWithState` that disposing has to cleanup any remaining `addState(...)` if you think there is any doubt and at best check all implementation of `IObjectWithState` in the platform.

I will add some vague wording.  There are no "working" implementations of IObjectWithState besides the ones I touched.

```java
        /**
	 * Adds state to this object.
	 *
	 * Usually, consumers will call <code>removeState()</code> with a matching id in
	 * the same lifecycle of IObjectWithState. However, this behavior is not
	 * guaranteed. Implementors should not rely on <code>removeState()</code> for
	 * resource disposal. The recommended practice is to free resources associated
	 * with non-removed states in some kind of dispose() method.
	 *
	 * @see AbstractHandlerWithState
	 *
	 * @param id    The identifier indicating the type of state being added; must
	 *              not be <code>null</code>.
	 * @param state The new state to add to this object; must not be
	 *              <code>null</code>.
	 */
	public void addState(String id, State state);
```

basilevs commented:
I have considered to track handler removals with `IEclipseContext.runAndTrack()` like suggested in https://github.com/eclipse-platform/eclipse.platform.ui/issues/1261 but as there are multiple contexts involved, I'm afraid this would cause listener leak.

basilevs committed:
Document occasional failure to invoke removeState() #925

Use of weak references in
org.eclipse.e4.core.commands.internal.HandlerServiceHandler may result
in occasional failure to call removeState() on an underlying handler
after use. We document this in IObjectWithState, until a better solution
is found.

iloveeclipse commented:
Let's merge that for the next build. Any objections?

laeubi commented:
> Let's merge that for the next build. Any objections?

go for it...

iloveeclipse commented:
@basilevs : thanks for a fast workaround.
